File: week2_capable/cleanup_duplicates.rb
#!/usr/bin/env ruby
# Cleanup duplicate rooms in world.db

require 'sqlite3'

DB_PATH = File.expand_path('.boukensha/world.db', __dir__)

unless File.exist?(DB_PATH)
  puts "❌ Database not found: #{DB_PATH}"
  exit 1
end

puts "🔍 Scanning for duplicate rooms in #{DB_PATH}..."

conn = SQLite3::Database.new(DB_PATH)
conn.results_as_hash = true
conn.execute("PRAGMA busy_timeout = 10000")

# Find duplicates
duplicates = conn.execute(<<~SQL)
  SELECT name, GROUP_CONCAT(id, ',') as ids, COUNT(*) as count
  FROM rooms
  GROUP BY name
  HAVING COUNT(*) > 1
  ORDER BY count DESC
SQL

if duplicates.empty?
  puts "✓ No duplicates found!"
  conn.close
  exit 0
end

puts "\n📊 Found #{duplicates.length} duplicate room names:\n"
duplicates.each do |row|
  puts "  • #{row['name']} (#{row['count']} copies)"
end

total_merged = 0

puts "\n🔗 Merging duplicates...\n"

duplicates.each do |group|
  name = group['name']
  room_ids = group['ids'].split(',').map(&:strip)

  next if room_ids.length < 2

  primary_id = room_ids[0]
  secondary_ids = room_ids[1..-1]

  puts "  📍 #{name}"
  puts "     Primary: #{primary_id}"
  puts "     Secondary: #{secondary_ids.join(', ')}"

  # For each secondary room
  secondary_ids.each do |secondary_id|
    # Redirect all exits pointing to secondary to primary
    updated = conn.execute(
      "UPDATE exits SET target_room_id = ? WHERE target_room_id = ?",
      [primary_id, secondary_id]
    )

    # Delete the secondary room
    conn.execute("DELETE FROM rooms WHERE id = ?", [secondary_id])

    total_merged += 1
    puts "     ✓ Deleted #{secondary_id}"
  end

  puts ""
end

conn.close

puts "✅ Done! Merged #{total_merged} duplicate room entries."
puts "   Map will show clean duplicates on next refresh."

File: week2_capable/cleanup_duplicates.py
#!/usr/bin/env python3
"""Cleanup duplicate rooms in world.db"""

import sqlite3
import sys
from pathlib import Path

DB_PATH = Path(__file__).parent / '.boukensha' / 'world.db'

if not DB_PATH.exists():
    print(f"❌ Database not found: {DB_PATH}")
    sys.exit(1)

print(f"🔍 Scanning for duplicate rooms in {DB_PATH}...")

conn = sqlite3.connect(str(DB_PATH))
conn.row_factory = sqlite3.Row
conn.execute("PRAGMA busy_timeout = 10000")

# Find duplicates
cursor = conn.execute("""
    SELECT name, GROUP_CONCAT(id, ',') as ids, COUNT(*) as count
    FROM rooms
    GROUP BY name
    HAVING COUNT(*) > 1
    ORDER BY count DESC
""")

duplicates = cursor.fetchall()

if not duplicates:
    print("✓ No duplicates found!")
    conn.close()
    sys.exit(0)

print(f"\n📊 Found {len(duplicates)} duplicate room names:\n")
for row in duplicates:
    print(f"  • {row['name']} ({row['count']} copies)")

total_merged = 0

print("\n🔗 Merging duplicates...\n")

for group in duplicates:
    name = group['name']
    room_ids = [rid.strip() for rid in group['ids'].split(',')]

    if len(room_ids) < 2:
        continue

    primary_id = room_ids[0]
    secondary_ids = room_ids[1:]

    print(f"  📍 {name}")
    print(f"     Primary: {primary_id}")
    print(f"     Secondary: {', '.join(secondary_ids)}")

    # For each secondary room
    for secondary_id in secondary_ids:
        # Redirect all exits pointing to secondary to primary
        conn.execute(
            "UPDATE exits SET target_room_id = ? WHERE target_room_id = ?",
            (primary_id, secondary_id)
        )

        # Delete the secondary room
        conn.execute("DELETE FROM rooms WHERE id = ?", (secondary_id,))

        total_merged += 1
        print(f"     ✓ Deleted {secondary_id}")

    print()

conn.commit()
conn.close()

print(f"✅ Done! Merged {total_merged} duplicate room entries.")
print("   Refresh the map to see the cleaned up rooms.")
